pkg/controllers/provisioningv2/rke2/unmanaged/controller.go

```diff
@@ -359,11 +372,43 @@ func (h *handler) onUnmanagedMachineOnRemove(key string, customMachine *rkev1.Cu
 	if err != nil {
 		return customMachine, err
 	}
-	err = clientset.CoreV1().Nodes().Delete(context.TODO(), machine.Status.NodeRef.Name, metav1.DeleteOptions{})
-	if apierror.IsNotFound(err) {
-		return customMachine, nil
+
+	removeAnnotation := "etcd." + rancherruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion) + ".cattle.io/remove"
+	removedNodeNameAnnotation := "etcd." + rancherruntime.GetRuntimeCommand(cluster.Spec.KubernetesVersion) + ".cattle.io/removed-node-name"
+
+	node, err := clientset.CoreV1().Nodes().Get(context.TODO(), machine.Status.NodeRef.Name, metav1.GetOptions{})
+	if err != nil {
+		if apierror.IsNotFound(err) {
+			return customMachine, nil
+		}
+		return customMachine, err
+	}
+
+	if val, ok := node.Annotations[removeAnnotation]; ok {
+		// check val to see if it's true, if not, continue
+		if val == "true" {
+			// check the status of the removal
+			if removedNodeName, ok := node.Annotations[removedNodeNameAnnotation]; ok {
+				// There is the possibility the annotation is defined, but empty.
+				if removedNodeName != "" {
+					return customMachine, nil
+				}
+			}
+		}
 	}
-	return customMachine, err
+	// The remove annotation has not been set to true, so we'll go ahead and set it on the node.
+	err = retry.RetryOnConflict(retry.DefaultRetry,
+		func() error {
+			node.Annotations[removeAnnotation] = "true"
+			node, err = clientset.CoreV1().Nodes().Update(context.TODO(), node, metav1.UpdateOptions{})
+			return err
+		})
+	if err != nil {
+		// there was an error updating the node
+		return customMachine, err
+	}
+
+	return customMachine, fmt.Errorf("waiting for etcd member removal")
```

Comment:
This code is very similar to the code in the `machineprovision` controller. Can this code be extracted to a function/method?

Reply:
I just consolidated this into its own package.